### PR TITLE
add guidance for bytes acked and clarification for segments_acked

### LINKED ISSUE #67: segments_acked rather than bytes_acked; cwnd in segments
[Yoshi said](https://mailarchive.ietf.org/arch/msg/tcpm/W-t1q8q6Q0Ofjcp0iXOgmY18mr0):
>8: Section 4.3  I am not very sure why segments_acked is used rather than byted_acked here. What is the benefit of it?
    How do we calculate when the acks are split? 
   Also, I think it should be clarified that cwnd is expressed in segments here in this case.

### PR DESCRIPTION
**Pull request by goelvidhi**
Closes #67

### CODE REVIEW

File: draft-ietf-tcpm-rfc8312bis.md [hunk 1]
```diff
@@ -376,6 +376,10 @@ during the different stages of the CUBIC congestion controller.
 
 The unit of all window sizes in this document is segments of the
 maximum segment size (MSS), and the unit of all times is seconds.
+Implementations can use bytes to express window sizes which would
```
Comment (larseggert) [inline draft-ietf-tcpm-rfc8312bis.md:379]:
```suggestion
Implementations can use bytes to express window sizes, which would
```

File: draft-ietf-tcpm-rfc8312bis.md [hunk 2]
```diff
@@ -446,7 +450,9 @@ An estimate for the congestion window in segments in the AIMD-friendly
 region, that is, an estimate for the congestion window of AIMD TCP.
 
 *segments_acked*:
-Number of segments acked when an ACK is received.
+Number of MSS-sized segments acked when an ACK is received. This
+number can be a decimal less than 1 when an ACK acknowledges a
+segment smaller than MSS.
```
Comment (larseggert) [inline draft-ietf-tcpm-rfc8312bis.md:455]: What if an ACK ACKs 1.5 MSS-sized segments? Is it 1 or 1.5 then?

**Review by larseggert — CHANGES_REQUESTED** [review]

**Review by lisongxu — APPROVED** [review]
Thank you!

**Comment by larseggert** [comment]
See my comment in the review - is this always a decimal or only when <1?

**Review by nsdyoshi — APPROVED** [review]

File: draft-ietf-tcpm-rfc8312bis.md (hunk 2 above)
Comment (junhochoi) [reply draft-ietf-tcpm-rfc8312bis.md]: I think it's 1.5. Also another question - does it include SACK'd size?

Comment (larseggert) [reply draft-ietf-tcpm-rfc8312bis.md]: @goelvidhi?

Comment (nsdyoshi) [reply draft-ietf-tcpm-rfc8312bis.md]: I think including SACKed bytes is an interesting question. I am thinking that if TCP receives SACKed packets, it means it's in recovery phase.

In my understanding, CUBIC doesn't touch the behavior in recovery phase. w_est will be initialized when TCP goes into congestion avoidance phase, which means after recovery phase has finished. So, I am thinking that we don't need to think about whether we should include SACKed bytes or not. 
But, I might overlook something, please let me know if there's situations where CUBIC should handle SACKed packets.

**Comment by goelvidhi** [comment]
> See my comment in the review - is this always a decimal or only when <1?

It is decimal always. Although most implementations don't do decimal computations and simply wait for one full congestion window to be acknowledged before incrementing the cwnd by alpha * 1 (segment)

**Comment by goelvidhi** [comment]
I can't see Junho's and Yoshi's comment about SACK'ed bytes here.. was it deleted?

**Comment by goelvidhi** [comment]
For my sanity, I will paste it from my email,
Junho said,
> I think it's 1.5. Also another question - does it include SACK'd size?

Yoshi replied,
>  think including SACKed bytes is an interesting question. I am thinking that if TCP receives SACKed packets, it means it's in recovery phase.
In my understanding, CUBIC doesn't touch the behavior in recovery phase. w_est will be initialized when TCP goes into congestion avoidance phase, which means after recovery phase has finished. So, I am thinking that we don't need to think about whether we should include SACKed bytes or not.
But, I might overlook something, please let me know if there's situations where CUBIC should handle SACKed packets.


I agree with Yoshi, we don't talk about recovery phase in detail but we allow an implementation to use a recovery algorithm
> For both packet loss
   and congestion detection through ECN, the sender MAY employ a Fast
   Recovery algorithm to gradually adjust the congestion window to its
   new reduced _ssthresh_ value.

**Comment by junhochoi** [comment]
> For my sanity, I will paste it from my email,
> Junho said,
> 
> > I think it's 1.5. Also another question - does it include SACK'd size?
> 
> Yoshi replied,
> 
> > think including SACKed bytes is an interesting question. I am thinking that if TCP receives SACKed packets, it means it's in recovery phase.
> > In my understanding, CUBIC doesn't touch the behavior in recovery phase. w_est will be initialized when TCP goes into congestion avoidance phase, which means after recovery phase has finished. So, I am thinking that we don't need to think about whether we should include SACKed bytes or not.
> > But, I might overlook something, please let me know if there's situations where CUBIC should handle SACKed packets.
> 
> I agree with Yoshi, we don't talk about recovery phase in detail but we allow an implementation to use a recovery algorithm
> 
> > For both packet loss
> > and congestion detection through ECN, the sender MAY employ a Fast
> > Recovery algorithm to gradually adjust the congestion window to its
> > new reduced _ssthresh_ value.

For example, when the sender received an ACK which doesn't move snd.una but one SACK record (dupack). Similar thing can happen with RACK's reordering threshold I think.

**Comment by nsdyoshi** [comment]
I see. yes, when we see small amount of re-ordering, you might receive SACKed packets.
In this case, I think you could include SACKed packets, but when it's ACKed accumulatively later, you will need to make sure you won't double count them.

**Comment by goelvidhi** [comment]
Aren't SACK'ed segments (before congestion event) covered in SACK RFC6675?
I would personally refrain from adding any SACK related details to Cubic or any specific congestion control spec.

**Comment by junhochoi** [comment]
@goelvidhi @nsdyoshi thanks for insights!

I asked about this because segments_acked is defined as "acked size in segment(or bytes)" and "acked" doesn't only means moving snd.una with SACK. Also QUIC RFC9002 "acked packet" always includes SACK'ed ones. So I thought the implementer might be confused what to use.

After thinking a little more.. SACK is optional as far as I understand, I think it might be good to leave as is.

**Comment by larseggert** [comment]
The comments didn't get deleted, they are inside my PR review. Yes, this is confusing.

**Comment by larseggert** [comment]
Should we make it clearer that this is always decimal, but stacks can be approximate in their math?

**Comment by goelvidhi** [comment]
@larseggert, how does this look,

Number of MSS-sized segments acked when an ACK is received. This
 number can be a decimal and it is less than 1 when an ACK acknowledges a
 segment smaller than the MSS.

**Comment by larseggert** [comment]
I'd maybe say "...is a decimal and can therefore take on non-integer values when an ACK acknowledges data that is not MSS-sized. Specifically, it can be less than 1 when an ACK acknowledges a segment smaller than the MSS..." or something like that?

**Comment by goelvidhi** [comment]
ok, I took your suggestion and made some minor edits. Please take a look.

For some reason, the inclusive check is getting cancelled immediately after I push.

**Comment by goelvidhi** [comment]
@larseggert, this PR is ready to go once you approve it. Not sure if we need to fix the inclusivity check before merging.

**Comment by goelvidhi** [comment]
@larseggert, should we merge this PR?

File: draft-ietf-tcpm-rfc8312bis.md [hunk 3]
```diff
@@ -446,7 +450,10 @@ An estimate for the congestion window in segments in the AIMD-friendly
 region, that is, an estimate for the congestion window of AIMD TCP.
 
 *segments_acked*:
-Number of segments acked when an ACK is received.
+Number of MSS-sized segments acked when an ACK is received. This
+number will be a non-integer value when an ACK acknowledges data
```
Comment (larseggert) [inline draft-ietf-tcpm-rfc8312bis.md:454]:
```suggestion
number will be a decimal value when an ACK acknowledges an amount of data
```

**Review by larseggert — CHANGES_REQUESTED** [review]
"Non-integer" is less accurate

**Pull request merged**